### ISSUE TITLE
Exception while decoding Base64 value with escaped `=` character

### ISSUE THREAD

**Issue by ViToni** [FasterXML/jackson-databind#2090]
While trying to deserialize a Base64 value I get this exception
```
com.fasterxml.jackson.databind.JsonMappingException: Failed to decode VALUE_STRING as base64 (MIME-NO-LINEFEEDS): Unexpected padding character ('=') as character #3 of 4-char base64 unit: padding only legal as 3rd or 4th character
```
which doesn't seem to make much sense.

The `ObjectMapper`  (2.8.8) is configured as follows:
```
        final ObjectMapper mapper = new ObjectMapper()
                .setSerializationInclusion(ALWAYS)
                .setBase64Variant(Base64Variants.MIME_NO_LINEFEEDS);
```

The values for the JSON are generated randomly, in this case the JSON was:
```
{"str":"","bytes":"Sp3VsWm7NEs1R0hAPumKiUarf+0iLEdWD2QmTxDmbxGDbzMSv0/04VTgTfARaGJTc1r6+jbTDLC/\r\n1tu1c3ZW5nEqBU3gOLdey2gvKaF+AukF+rTJbAPMl5xAr/RPetqzZ24wLRYc1TcfzUpEzOJOtUHK\r\nPq0Z6OBob5yQNZoFt0y56HoHYsFsW3p45Zk/jrOermSMXoXGsfvkKf53XVlPfUBwNXz0NdqmKwV4\r\nTwDl8ZOWIOi9bchU2QumijA38pek2zF+U7S+3igjfY/05Sv8LYiH2ieCljSh1xPC4NjmosLo7Ovp\r\nvZrkBHgqZnUaT1l82dn8HejCvTOiFK8+h5bpr88/tMvclQhx5feV7J/4yoYeo+dwuANfzKfQgmmd\r\np625dwSfDBVr0xNIYYbnyPYv7LNer8x+PWZ0h28hFwhWC1HPvFijHvfM+ST0d8O78txLNyyhhbtd\r\nhB/FIJWcyRGX00cP9o7jOlN2HWBP4/V25dMOlDHHZAUCUkq0qTF2baC1VFO61QAuX4Z+F4KgxkUP\r\nZN5bT9NhA+9n4cIy1rlcdH/vLXN8SmEht7Z/xgPQ/2kyEmuM1A+f1nO8Zr7NzsI4IBc++21HcaAl\r\n1rFphIwaxcNav87hv7LCMu2DtrwJUTsSStm16rBIw+cVy2xHxSZ+VylxezY0lpfC7qAyEE61za0Q\r\n2BWA9jJxEkzNy5kaZ2GbNshNTinZEp8aT/x/lxgJ0XFygbi7vrGbew+Qq4C2lpn7X0yIyc326DEj\r\niChy7RK0eSwlrFVw9opCPpDjv77UVzlcjjA14ywi2D3l8IAOvVq70odbTkT4RGjs/33wflSMcq3C\r\nVALQ6FgXZOWu+CyblcuG8UTmAdNDWLMwigUiPyJuZauTuIwWM9mn/8Ehj+ftw9v39Y9dSDef4ybG\r\nQSWQBLcb/ytUHp1qb6Cvq9se0Si/AsFycOtUXHDF40i8zqXLHGJi68YwQBINM1OCF8gMSGEH0Czk\r\n4YDWIGmHZuDz+XYySWtCdIEqakzryPlaJ8QOTRxs6/irlviqB6y77w+9Uc704K62GWsip+NQIz+I\r\nOxuu9/R38EuptSvznNd6qk1+jECp4VhZ76j+PXL08jPJlrA/thjCjZTmZxdLj+sNxmlLdBaCG9HD\r\nhxUs5xq1D/hWAwcM5ZkJYatDCiGd/E0alq8IhzQz7aXnFojRoV6L7wtfWerFU/Y180msGNQ+noC+\r\nnnuJlspvMAIL/bly1996wfrXGYYEzMIgJ1m01pT4HVfENPY2fLssrmW/G8udnUmJE2a7rXG2StQw\r\nqL30tRMxzg96sSzZ7dI3eNV+VUKLQ55r5eQElte7H9m864ytqTXCM3QH0+fJAmb872Kz7NMJsg\u003d\u003d","map":{},"list":null}
```
 
Background:
I need send data from A to B (and back), where A is fixed to use GSON and B to jackson
Now I'm trying to get a confguration for A & B which works for all cases.
To be sure I'm setting up something like a test suite to cover as many scenarios as possible.
As I'm using the Base64 encoder from the JDK (1.8.172) I thought to be on the safe side..

As reference the tested class:
```
package dto;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DataContainer {
    private String str;
    private byte[] bytes;
    private Map<String, String> map;
    private List<String> list;

    protected DataContainer() {
        // only used by deserializer
    }

    public DataContainer(
            final String str,
            final byte[] bytes,
            Map<String, String> map,
            List<String> list
    ) {
        this.str = str;
        this.bytes = bytes == null ? null : Arrays.copyOf(bytes, bytes.length);
        this.map = map == null ? null : Collections.unmodifiableMap(new HashMap<>(map));
        this.list = list == null ? null : Collections.unmodifiableList(new ArrayList<>(list));
    }

    public String getStr() {
        return str;
    }

    public byte[] getBytes() {
        return bytes;
    }

    public Map<String, String> getMap() {
        return map;
    }

    public List<String> getList() {
        return list;
    }

    @Override
    public int hashCode() {
        int tmpHashCode = 37;
        tmpHashCode = (tmpHashCode + (str == null ? 0 : str.hashCode())) * 17;
        tmpHashCode = (tmpHashCode + Arrays.hashCode(bytes)) * 17;
        tmpHashCode = (tmpHashCode + (map == null ? 0 : map.hashCode() )) * 17;
        tmpHashCode = (tmpHashCode + (list == null ? 0 : list.hashCode() )) * 17;

        return tmpHashCode;
    }

    @Override
    public boolean equals(final Object obj) {
        if (this == obj) {
            return true;
        }
        if (null == obj) {
            return false;
        }

        if (getClass() == obj.getClass()) {
            final DataContainer dc = (DataContainer) obj;
            return
                    ((str == null && dc.str == null) || (str != null && str.equals(dc.str))) &&
                    Arrays.equals(bytes, dc.bytes) &&
                    ((map == null && dc.map == null) || (map != null && map.equals(dc.map))) &&
                    ((list == null && dc.list == null) || (list != null && str.equals(dc.list)));
        }
        return false;
    }
}
```

and the used converter (jackson 2.8.8):
```
package dto;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.ALWAYS;

import java.io.IOException;

import com.fasterxml.jackson.core.Base64Variants;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectReader;
import com.fasterxml.jackson.databind.ObjectWriter;

public class JacksonConverter
    implements GenericConverter<DataContainer>{

    private final ObjectReader textReader;
    private final ObjectWriter textWriter;

    public JacksonConverter() {
        final ObjectMapper mapper = new ObjectMapper()
                .setSerializationInclusion(ALWAYS)
                .setBase64Variant(Base64Variants.MIME_NO_LINEFEEDS);

        this.textReader = mapper.readerFor(DataContainer.class);
        this.textWriter = mapper.writerFor(DataContainer.class);
    }

    public DataContainer fromText(final String text) throws IOException {
        return textReader.readValue(text);
    }

    public String toText(final DataContainer dataContainer) throws IOException {
        return textWriter.writeValueAsString(dataContainer);
    }

}
```

and the other used converter (GSON 2.2.5):
```
package dto;

import java.io.IOException;
import java.lang.reflect.Type;
import java.util.Base64;
import java.util.Objects;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;
import com.google.gson.JsonPrimitive;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.JsonSyntaxException;

public class GsonConverter
    implements GenericConverter<DataContainer>{

    private final Gson converter = new GsonBuilder() //
            .serializeNulls()//
            .registerTypeHierarchyAdapter(byte[].class, new ByteArrayToBase64TypeAdapter()) //
            .create();

    public String toText(final DataContainer object) throws IOException {
        Objects.requireNonNull(object, "object");
        return converter.toJson(object);
    }

    public DataContainer fromText(final String jsonString) throws IOException {
        return fromText(jsonString, DataContainer.class);
    }

    protected <T> T fromText(final String jsonString, final Class<T> classOfObject) throws JsonSyntaxException {
        Objects.requireNonNull(jsonString, "jsonString");
        Objects.requireNonNull(classOfObject, "classOfObject");
        return converter.fromJson(jsonString, classOfObject);
    }

    private static class ByteArrayToBase64TypeAdapter
        implements JsonSerializer<byte[]>, JsonDeserializer<byte[]> {

        private final Base64.Decoder decoder = Base64.getMimeDecoder();
        private final Base64.Encoder encoder = Base64.getMimeEncoder();

        public byte[] deserialize(final JsonElement json, final Type typeOfT, final JsonDeserializationContext context) throws JsonParseException {
            final byte[] bytes = decoder.decode(json.getAsString());
            return bytes;
        }

        public JsonElement serialize(final byte[] src, final Type typeOfSrc, JsonSerializationContext context) {
            final String bytesAsBase64Str = encoder.encodeToString(src);
            return new JsonPrimitive(bytesAsBase64Str);
        }
    }
}
```

**Comment by ViToni** [comment]
Turns out, the issues was in the GsonBuilder setup.

HTML escaping has to be disabled with ` disableHtmlEscaping()` to avoid conversion of `=` to `\u003d` which is not understood by jackson.

**Comment by cowtowncoder** [comment]
@ViToni Thank you for debugging this. I _think_ there might actually be something Jackson should do better, if I understand things correctly. JSON escaping should be legal, even if ideally not used for '=' character.

**Comment by ViToni** [comment]
@cowtowncoder You're welcome :)

Maybe while working on the encoding...
As far I understood MIME mandates the usage of CRLF as line break. The ouput from jackson seems to use LF only which might not fit to the RFC. (I could create an own issue for that if found to be useful.)

**Comment by cowtowncoder** [comment]
Recreated original issue as

https://github.com/FasterXML/jackson-core/issues/477

**Comment by sksamal** [comment]
Is this issue fixed.? I am getting the same issue with jackson-databind 2.9.11

**Comment by cowtowncoder** [comment]
@sksamal Yes, the linked-to issue has been fixed.

If you have remaining issues with base64 decoding, you may want to file a new issue under `jackson-core` (that's where Base64 decoding is handled) but make sure to first test with a newer version of Jackson: 2.9.x is no longer supported so even if there was a problem, no fixes would be applied to 2.9 branch.
So reproduction against 2.11.4 or 2.12.3 would make sense.
